build-support/bin/generate_builtin_lockfiles.py

```diff
@@ -0,0 +1,309 @@
+from pants.backend.helm.subsystems.post_renderer import HelmPostRendererSubsystem
+from pants.backend.java.lint.google_java_format.subsystem import GoogleJavaFormatSubsystem
+from pants.backend.java.subsystems.junit import JUnit
+from pants.backend.kotlin.lint.ktlint.subsystem import KtlintSubsystem
+from pants.backend.python.goals.coverage_py import CoverageSubsystem
+from pants.backend.python.lint.add_trailing_comma.subsystem import AddTrailingComma
+from pants.backend.python.lint.autoflake.subsystem import Autoflake
+from pants.backend.python.lint.bandit.subsystem import Bandit
+from pants.backend.python.lint.black.subsystem import Black
+from pants.backend.python.lint.docformatter.subsystem import Docformatter
+from pants.backend.python.lint.flake8.subsystem import Flake8
+from pants.backend.python.lint.isort.subsystem import Isort
+from pants.backend.python.lint.pydocstyle.subsystem import Pydocstyle
+from pants.backend.python.lint.pylint.subsystem import Pylint
+from pants.backend.python.lint.pyupgrade.subsystem import PyUpgrade
+from pants.backend.python.lint.ruff.subsystem import Ruff
+from pants.backend.python.lint.yapf.subsystem import Yapf
+from pants.backend.python.packaging.pyoxidizer.subsystem import PyOxidizer
+from pants.backend.python.subsystems.debugpy import DebugPy
+from pants.backend.python.subsystems.ipython import IPython
+from pants.backend.python.subsystems.lambdex import Lambdex
+from pants.backend.python.subsystems.pytest import PyTest
+from pants.backend.python.subsystems.python_tool_base import PythonToolRequirementsBase
+from pants.backend.python.subsystems.setuptools import Setuptools
+from pants.backend.python.subsystems.setuptools_scm import SetuptoolsSCM
+from pants.backend.python.subsystems.twine import TwineSubsystem
+from pants.backend.python.typecheck.mypy.subsystem import MyPy
+from pants.backend.scala.lint.scalafmt.subsystem import ScalafmtSubsystem
+from pants.backend.scala.subsystems.scalatest import Scalatest
+from pants.backend.terraform.dependency_inference import TerraformHcl2Parser
+from pants.backend.tools.yamllint.subsystem import Yamllint
+from pants.base.build_environment import get_buildroot
+from pants.jvm.resolve.jvm_tool import JvmToolBase
+from pants.util.contextutil import temporary_dir
+from pants.util.dirutil import touch
+
+logger = logging.getLogger(__name__)
+
+
+ToolBaseT = TypeVar("ToolBaseT")
+
+
+@dataclass
+class Tool(Generic[ToolBaseT]):
+    cls: Type[ToolBaseT]
+    backend: str
+
+    @property
+    def name(self) -> str:
+        return cast(str, self.cls.options_scope)  # type: ignore[attr-defined]
+
+    @property
+    def resolve(self) -> str:
+        return self.name
+
+    @property
+    def lockfile_name(self) -> str:
+        return f"{self.name}.lock"
+
+
+@dataclass
+class PythonTool(Tool[PythonToolRequirementsBase]):
+    ...
+
+
+@dataclass
+class JvmTool(Tool[JvmToolBase]):
+    ...
+
+
+python_interpreter_constraints = "CPython>=3.7,<3.12"
```

Comment:
The point of the Pex / Lambdex `<3.12` restriction is not to guard against 4.x, its actually to guard against breaking changes to Python build related stdlib (distutils) in 3.next; i.e. actually 3.12 for real.

Reply:
That said, having 1 value here seems like a luck circumstance unless it's true Pants refuses to support shipping tools that don't work with Python 3.7 - for example. Presumably folks might want to ship a hot of the press tool that does not support 3.7 since 3.7 is dead or just about so, etc.